fuzzers/main.py

```diff
@@ -0,0 +1,642 @@
+
+def pause_exit_spot(pause_evt: threading.Event, resume_evt: threading.Event, exit_evt: threading.Event) -> None:
+    """
+    Poll the events, returning True if `exit_evt` was set and pausing as requested by `pause_evt`
+    and `resume_evt`
+    """
+
+    resume_evt.clear()
+    if pause_evt.is_set():
+        while not resume_evt.wait(timeout=1):
+            if exit_evt.is_set():
+                return True
+        pause_evt.clear()
+    return exit_evt.is_set()
+
+def kill_fuzzers(bucket: gcs.Bucket, fuzzers: typing.Iterable[FuzzProcess]):
+    for f in fuzzers:
+        try:
+            f.signal(signal.SIGTERM)
+        except ProcessLookupError:
+            print(f"Failed looking up process {f.proc.pid}")
+    time.sleep(5)
+    for f in fuzzers:
+        try:
+            f.signal(signal.SIGKILL)
+        except ProcessLookupError:
+            print(f"Failed looking up process {f.proc.pid}")
+    for f in fuzzers:
+        bucket.blob(f"logs/{f.log_path}").upload_from_filename(LOGS_DIR / f.log_path)
+
+def run_fuzzers(gcs_client: gcs.Client, pause_evt: threading.Event, resume_evt: threading.Event, exit_evt: threading.Event):
+    """
+    Run all the fuzzers until `exit_evt` gets triggered, pausing and resuming
+    them based on `pause_evt` and `resume_evt`.
+    """
+
+    bucket = gcs_client.bucket(GCS_BUCKET)
+
+    repo = Repository(REPO_DIR, REPO_URL)
+    repo.clone_if_need_be()
+
+    corpus = Corpus(CORPUS_DIR, bucket)
+
+    while True:
+        if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+            return
+
+        sync_log_files = []
+        date = datetime.datetime.now().strftime('%Y-%m-%d')
+
+        # Read the configuration from the repository
+        cfg = repo.latest_config('master')
+        cfg_for = {b['name']: repo.latest_config(b['name']) for b in cfg['branch']}
+
+        # Figure out which targets we want to run
+        branches = random_weighted(cfg['branch'], NUM_FUZZERS)
+        targets = [random_weighted(cfg_for[b['name']]['target'], 1)[0] for b in branches]
+
+        # Synchronize the relevant corpuses
+        corpus.stop_synchronizing()
+        corpus.update()
+        for t in set((t['crate'], t['runner']) for t in targets):
+            log_path = pathlib.Path('sync') / date / t[0] / t[1] / str(uuid.uuid4())
+            utils.mkdirs((LOGS_DIR / log_path).parent)
+            corpus.synchronize(t[0], t[1], open(LOGS_DIR / log_path, 'a'))
+            sync_log_files.append(log_path)
+            if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+                return
+
+        # Initialize the fuzzers
+        fuzzers = []
+        for (b, t) in zip(branches, targets):
+            worktree = repo.worktree(b['name'])
+            log_path = pathlib.Path('fuzz') / date / t['crate'] / t['runner'] / str(uuid.uuid4())
+            utils.mkdirs((LOGS_DIR / log_path).parent)
+            log_file = LOGS_DIR / log_path
```

Comment:
Or swap the two lines and you won’t need to calculate the path twice.

Reply:
Oh right I refactored `log_file` as a path instead of an actual file and forgot to adjust that :)